### ISSUE TITLE
`--without` and “Gems in at least one non-excluded group will still be installed” at odds with PLOS

### ISSUE THREAD

**Issue by yb66** [rubygems/rubygems#3329]
Evening,

This is from [the docs on groups](https://bundler.io/guides/groups.html):

> Gems in at least one non-excluded group will still be installed

Forgive me for being somewhat philosophical, but I struggle to think of an occasion where I know the _context_ and _explicitly_ wish to exclude something, and it is then included because of some part of the context. If I say I don't eat meat, I really don't care if you serve me Indian food or Chinese food with meat in it, I don't want it. I don't care if it comes with veg or fruit, _I don't want it_.

This extends to tech, the authorisation for accessing files on a filesystem or tables in a database - it doesn't matter if you're in the Sales team and they have access, if you've been explicitly excluded then you don't get access.

Yet Bundler is going against explicitly requested exclusion and given what I've written above, goes against PLOS. If I have the following Gemfile and the `--without development` my Gemfile.lock will include coffee-script (expected) and my gems will include coffee-script (completely unexpected). I confirmed this is the case both Bundler v2.0.1 and v1.16.1.

```

group :web, :production do
  gem "sinatra"
  gem "haml"
end

group :assets, :development do
  gem "coffee-script"
end
```

This has knock on effects as I try to find a way around it. My specific need right now is to create a template Docker install that runs `bundle install --without`… so this is stopping me. My next move was to use ENV vars but as the Gemfile.lock needs to be updated I have to remember to run an install with the production ENV vars set to update the lock file or there'll be errors (there _are_ errors). I've considered creating a separate production lock file or not copying the lock file over (that seems really bad) or perhaps some convoluted wrapper script, and there are many other _workarounds_ I can use but it all boils down to, what I find to be, a strange choice.

I've looked through the docs and the source and can't find any explanation for this decision.

My 2/3 suggestions are:

1. An explicit exclusion is just that, **an exclusion** regardless of a gem's inclusion in another group.
2. The Gemfile.lock has separate sections for `production`/`deployment` and `development` (and possibly `test` or more).
3. Or, a `Gemfile.lock.production` etc is generated.

These suggestions do not, I believe, conflict with [the stated core goal of Bundler](https://github.com/bundler/bundler/issues/399#issuecomment-285525) 

> The entire point of bundler is a consistent set of gem versions

and [from the same conversation](https://github.com/bundler/bundler/issues/399#issuecomment-285398)

> Bundler's core function is to resolve the graph of gem dependencies, and that is impossible without every gemspec.

An entire graph _per environment_ would still be generated. The `--without` switch would simply not install gems. Everything would be the same except my surprise at finding extra gems installed. Either way, suggestion (1) is the main thing, not to install that which has been explicitly requested not to be.

I've no idea if this should be an RFC.

Thanks for your time.

Regards,
iain

**Comment by deivid-rodriguez** [comment]
Hi @yb66! Sorry for the really big delay.

Honestly, my expectations are the same as yours, and I'd say this is just a bug. But I might be missing something. I'll look more closely.

**Comment by deivid-rodriguez** [comment]
Oh, wait, I understand this now.

The gem is getting installed because you're not excluding the `assets` group, so since the gem is on a non-excluded group (`assets`), it's getting installed as part of the `assets` group. I believe there's 3 options here:

* Starting doing what you expect.
* Keep doing what we do.
* Starting throwing an error or warning in these "ambiguous" situations.

I'd lean towards throwing a warning in these ambiguous situations, and eventually fail.

For the time being, you can configure explicitly what you want `bundler` to do with gems in the `assets` group. You can probably either configure `bundle config set --local without assets` or remove the `assets` group altogether.

**Comment by yb66** [comment]
Hi @deivid-rodriguez 

> Sorry for the really big delay

Don't worry, you're 6 months quicker than the Bundler team were ;-P I appreciate the reply, sincerely, don't worry about it.

> Starting throwing an error or warning in these "ambiguous" situations.

I agree, some form of communication when this happens would be helpful. Not necessarily an error but knowledge is good. Managing a graph in your head isn't optimal :)

I can see the merits of "Keep doing what we do" in a pragmatic sense, of course, but Bundler and Rubygems are not your average programmer tools because they're not used exclusively by programmers. For instance, those who set up servers are going to want an explict deny to beat an implicit allow. It could end up as a security problem, the kind that explicit deny is designed to defeat.

Anyway, thanks for your help, whatever you decide is fine with me.

Regards,
iain

**Comment by deivid-rodriguez** [comment]
I essentially agree with your sentiment.

In my opinion, the biggest pain point here, is that bundler is giving incorrect information to the user. In the sample Gemfile you posted, with `without development` configured, it will print the following message to the screen:

> Gems in the group development were not installed.

Which is... not true. It should instead print

> Gems in the group development were not installed, except for "coffee-script-source" because it's also part of the non-excluded group assets.

With that bug fix in place, I think we reduce most of the pain caused by this issue, and most suprises.

In addition to that, you also have a point that an explicit exclusion should probably take precedence over an implicit inclusion. We can also consider changing this for the next major version.

**Comment by yb66** [comment]
I completely agree. How do you wish to proceed? I'm not sure the Bundler team will pay any attention to me but let me know if I should open another issue with them about the error message.

Regards,
iain

**Comment by deivid-rodriguez** [comment]
Yes, please, let's handle improving the error message as a separate issue.

Also, just to clarify, I'm part of the Bundler team, just like other maintainers who, like me, couldn't find time to dedicate to this issue earlier. We have limited availability and are no always able to respond to issues in a timely manner.
